src/Microsoft.Diagnostics.Monitoring.EventPipe/Counters/TraceEventExtensions.cs

```diff
@@ -37,20 +37,64 @@ internal record struct ProviderAndCounter(string ProviderName, string CounterNam
     internal static class TraceEventExtensions
     {
         private static Dictionary<ProviderAndCounter, CounterMetadata> counterMetadataByName = new();
+        private static Dictionary<int, CounterMetadata> counterMetadataById = new();
         private static HashSet<string> inactiveSharedSessions = new(StringComparer.OrdinalIgnoreCase);
 
-        // This assumes uniqueness of provider/counter combinations;
-        // this is currently a limitation (see https://github.com/dotnet/runtime/issues/93097 and https://github.com/dotnet/runtime/issues/93767)
-        public static CounterMetadata GetCounterMetadata(string providerName, string counterName, string meterTags = null, string instrumentTags = null, string scopeHash = null)
+        private static CounterMetadata AddCounterMetadata(string providerName, string counterName, int? id, string meterTags, string instrumentTags, string scopeHash)
         {
+            CounterMetadata metadata;
+            if (id.HasValue && counterMetadataById.TryGetValue(id.Value, out metadata))
+            {
+                return metadata;
+            }
+
+            // Its possible that we previously indexed this counter by name but it didn't have an ID at that point because we weren't
+            // listening to it then.
+            // Its also possible that we previously indexed a counter with the same name as this one but with different tags or scope hash.
             ProviderAndCounter providerAndCounter = new(providerName, counterName);
-            if (counterMetadataByName.TryGetValue(providerAndCounter, out CounterMetadata provider))
+            if (counterMetadataByName.TryGetValue(providerAndCounter, out metadata))
             {
-                return provider;
+                // we found a counter that matches the name, but it might not match everything
+                if (metadata.MeterTags == meterTags && metadata.InstrumentTags == instrumentTags && metadata.ScopeHash == scopeHash)
+                {
+                    // add the ID index if it didn't exist before
+                    if (id.HasValue)
+                    {
+                        counterMetadataById.TryAdd(id.Value, metadata);
+                    }
+                    return metadata;
+                }
             }
 
-            counterMetadataByName.Add(providerAndCounter, new CounterMetadata(providerName, counterName, meterTags, instrumentTags, scopeHash));
-            return counterMetadataByName[providerAndCounter];
+            // no pre-existing counter metadata was found, add a new one
+            metadata = new CounterMetadata(providerName, counterName, meterTags, instrumentTags, scopeHash);
+            if (id.HasValue)
+            {
+                counterMetadataById.TryAdd(id.Value, metadata);
+            }
+            counterMetadataByName.TryAdd(providerAndCounter, metadata);
+            return metadata;
```

Comment:
Just reading through all of this, I think everything makes sense - I think there might be one edge case here to call out (quite possible that I'm misreading this though, I haven't played with it). 

Previously, every time you tried to `GetCounterMetadata`, if there was a match, you got the existing metadata back, even though it may have been a name collision (we had no way to know, since we're just going off provider/counter names). Otherwise, we would create a new entry for the provider/counter, and return that new metadata.

Now, in `AddCounterMetadata` it looks like if there's a collision, we check if the tags/scope match, and if not, we don't return the existing metadata. Instead, we create new metadata...but it's possible that both 73 and 75 here fail (because that provider/counter may already exist with those different tags/scopes). However, we return that new metadata on line 76, even if our dictionary still points to the existing entry that has different tags/scope.

I guess I'm not sure if this has any real ramifications, but if I'm reading it correctly that seems like it could be misleading - we return this newly created metadata, even though every time we call `GetCounterMetadata` we're actually still going to be using the old one.

Reply:
Yeah, that was intentional though let me know if you expect it would be problematic for dotnet-monitor. Previously BeginInstrumentReportingPayload might contain a CounterMetadata object where the Tags/Hash differed from the data that was in the underlying EventPipe event. After this change BeginInstrumentReportingPayload always reports the same Tags/Hash as underlying event. If there was no ID and a name collision then the new CounterMetadata hits that case where it isn't cached and won't be seen again.

dotnet-counters doesn't actually care about the tags/hash reported in the BeginInstrumentReportingPayload but it seemed far less confusing to make the payload true to the underlying event data rather than coerce it to use a non-matching cached value.